trio/_core/_run.py

```diff
@@ -2374,13 +2376,17 @@ async def checkpoint_if_cancelled():
     task._cancel_points += 1
 
 
-if os.name == "nt":
+if sys.platform == "win32":
     from ._io_windows import WindowsIOManager as TheIOManager
     from ._generated_io_windows import *
-elif hasattr(select, "epoll"):
+elif sys.platform == "linux" or (not TYPE_CHECKING and hasattr(select, "epoll")):
     from ._io_epoll import EpollIOManager as TheIOManager
     from ._generated_io_epoll import *
-elif hasattr(select, "kqueue"):
+elif (
+    sys.platform == "darwin"
+    or sys.platform.startswith("freebsd")
+    or (not TYPE_CHECKING and hasattr(select, "epoll"))
+):
```

Comment:
Well... this check is gross and we should probably complain about it to typing-sig.

But having written it, I guess we might as well copy-paste it into `trio/_core/__init__.py` too?

Reply:
Oh heh and there's a copy/paste error here – you accidentally replaced `hasattr(select, "kqueue")` with `hasattr(select, "epoll")`.

Reply:
> Well... this check is gross and we should probably complain about it to typing-sig.

Do you mean that mypy should be able to tell wich platforms support epoll/kqueue?

> But having written it, I guess we might as well copy-paste it into `trio/_core/__init__.py` too?

Do you mean that this hack is better than the current `try/except ImportError` way in `trio/_core/__init__.py`?

> Oh heh and there's a copy/paste error here – you accidentally replaced `hasattr(select, "kqueue")` with `hasattr(select, "epoll")`.

Thanks, fixed.

Reply:
> Do you mean that mypy should be able to tell wich platforms support epoll/kqueue?

Maybe? I mean that having to copy-paste this complicated and not-even-terribly-correct code into multiple places is suboptimal, and maybe it could be avoided somehow. But yeah, teaching mypy to evaluate `hasattr` statements seems like one good way to do that... it already has a model of which attributes the `select` module has!

And heh, on further investigation, it looks like the way mypy's understanding of `select` works is:

```python
if sys.platform != "linux" and sys.platform != "win32":
```

https://github.com/python/typeshed/blob/66a9a4b5ce72aaf022b24706bf58254602ef02f8/stdlib/2and3/select.pyi#L36-L38

That is probably nicer than trying to enumerate all the BSD variations... so I guess maybe for now we should do:

```python
if sys.platform == "win32":
    ...
elif sys.platform == "linux" or hasattr(select, "epoll"):
    ...
elif TYPE_CHECKING or hasattr(select, "kqueue"):
    ...
else:
    raise NotImplementedError
```

> Do you mean that this hack is better than the current try/except ImportError way in trio/_core/__init__.py?

Yeah, b/c the way that's written right now, if you try to run mypy on macOS then it'll complain that `current_kqueue` and friends are undefined.

Reply:
Thanks, done. (Btw, note that you can use "Quote reply" to preserve the markdown formatting of the text you're quoting.)

Reply:
Any arm of the if that *might* be true (according to mypy's very limited static analysis) *must* type-check. That's the reason for all the `not TYPE_CHECKING and` in the original -- it basically is saying "I know you can't tell this hasattr() result when type-checking, so just ignore it".

Reply:
So I guess that means we should do something like this?

```python
if sys.platform == "win32":
    ...
elif sys.platform == "linux" or (not TYPE_CHECKING and hasattr(select, "epoll")):
    ...
elif TYPE_CHECKING or (not TYPE_CHECKING and hasattr(select, "kqueue")):
    ...
else:
    raise NotImplementedError
```

?

Reply:
Also raised this upstream here: https://github.com/python/mypy/issues/9042

Reply:
I don't think you need the `not TYPE_CHECKING` guard in the kqueue arm, because mypy can tell that `TYPE_CHECKING or ...` is true regardless of the `...`.

Reply:
Thanks, switched to this. Will probably help with multiplatform support even if right now I get the exact same errors